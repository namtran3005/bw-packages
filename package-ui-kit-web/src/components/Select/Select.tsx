import React, { useState } from 'react';
import styled from 'styled-components';
import { Select as MuiSelect } from '@material-ui/core';
import { SelectProps as MuiSelectProps } from '@material-ui/core/Select';
import { Colors } from '@bitwala-cryptobank-squad/package-theme';
import { ChevronDown } from '@bitwala-cryptobank-squad/package-icons';

type SelectProps = Omit<MuiSelectProps, 'variant'> & {
  disableUnderline?: boolean;
  isSelectOpen?: boolean;
};

interface SelectIconProps {
  isOpen: boolean;
}

const SelectComponent: React.FunctionComponent<SelectProps> = ({
  disableUnderline = true,
  ...props
}) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <MuiSelect
      disableUnderline={disableUnderline}
      IconComponent={() => <ArrowDownStyled isOpen={isOpen} />}
      onOpen={() => setIsOpen(true)}
      onClose={() => setIsOpen(false)}
      style={{
        background: isOpen
          ? Colors.get('white')
          : Colors.getWithAlpha('subtitleText', 0.1),
      }}
      MenuProps={{
        anchorOrigin: {
          vertical: 'bottom',
          horizontal: 'center',
        },
        transformOrigin: {
          vertical: 'top',
          horizontal: 'center',
        },
        getContentAnchorEl: null,
        PaperProps: {
          elevation: 0,
          style: {
            marginTop: '16px',
          },
        },
      }}
      {...props}
    />
  );
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const ArrowDownStyled = styled(({ isOpen, ...props }: SelectIconProps) => (
  <span {...props}>
    <ChevronDown />
  </span>
))`
  && {
    display: flex;
    position: absolute;
    right: 3px;
    justify-content: center;
    pointer-events: none;

    svg {
      width: 50%;
      height: 50%;
      margin-top: ${({ isOpen }) => (isOpen ? '-25%' : '25%')};
      transform: ${({ isOpen }) => (isOpen ? 'scaleY(-1)' : '')};
    }
  }
`;

const Select = styled(SelectComponent)`
  && {
    width: 100%;
    height: 48px;
    border-radius: 4px;
    margin: 6px 0px;
    font-family: Inter;
    font-size: 0.875rem;
    font-weight: 500;
    border: 1px solid
      ${({ ...props }) => (props.error ? Colors.get('error') : 'transparent')};
  }

  select {
    padding-left: 12px;
    padding-top: 0px;
    padding-bottom: 0px;
    height: 100%;
  }

  div {
    display: flex;
    align-items: center;
    padding-left: 12px;
    padding-top: 0px;
    padding-bottom: 0px;
    height: 100%;
  }

  svg {
    opacity: ${({ ...props }) => (props.disabled ? 0.5 : 1)};
  }
`;

export { Select, SelectProps };
