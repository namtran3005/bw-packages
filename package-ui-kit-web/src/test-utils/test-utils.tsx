import {
  render as rtlRender,
  waitFor,
  RenderOptions as RTLRenderOptions,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { WebAppThemeProvider } from '../theme';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';

type RenderOptions = RTLRenderOptions;

function renderWithTestHarness(
  ui: React.ReactElement,
  renderOptions: RenderOptions = {}
) {
  function Wrapper({ children }: any) {
    return (
      <MemoryRouter>
        <WebAppThemeProvider>{children}</WebAppThemeProvider>
      </MemoryRouter>
    );
  }

  return rtlRender(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}

// Use if you're in a hurry (better than nothing)
// be sure to verify no memory leak from implicit
// waitFor by adding these flags to jest: --runInBand --detectOpenHandles
async function renderAct(ui: React.ReactElement, options: RenderOptions = {}) {
  const view = renderWithTestHarness(ui, options);
  return await waitFor(() => view, {
    timeout: 5000,
    onTimeout: (e) => {
      // eslint-disable-next-line no-console
      console.error(e);
      return e;
    },
  });
}

export * from '@testing-library/react';

export { renderWithTestHarness, renderAct, userEvent };
