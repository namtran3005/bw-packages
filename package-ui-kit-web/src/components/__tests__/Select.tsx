import * as React from 'react';

import { renderWithTestHarness } from '../../test-utils/test-utils';
import { Select } from '../Select';

const defaultProps = {};

describe('Select', () => {
  it('should render input label', () => {
    const { baseElement } = renderWithTestHarness(<Select {...defaultProps} />);
    expect(baseElement).toMatchSnapshot();
  });
});
