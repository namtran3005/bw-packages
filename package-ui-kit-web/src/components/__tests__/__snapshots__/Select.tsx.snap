// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render input label 1`] = `
.c1.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  right: 3px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
}

.c1.c1 svg {
  width: 50%;
  height: 50%;
  margin-top: 25%;
}

.c0.c0 {
  width: 100%;
  height: 48px;
  border-radius: 4px;
  margin: 6px 0px;
  font-family: Inter;
  font-size: 0.875rem;
  font-weight: 500;
  border: 1px solid transparent;
}

.c0 select {
  padding-left: 12px;
  padding-top: 0px;
  padding-bottom: 0px;
  height: 100%;
}

.c0 div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 12px;
  padding-top: 0px;
  padding-bottom: 0px;
  height: 100%;
}

.c0 svg {
  opacity: 1;
}

<body>
  <div>
    <div
      class="MuiInputBase-root MuiInput-root c0"
      style="background: rgba(109, 109, 134, 0.1);"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby=" "
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        role="button"
        tabindex="0"
      >
        <span>
          â€‹
        </span>
      </div>
      <input
        type="hidden"
        value=""
      />
      <span
        class="c1"
      >
        <svg
          fill="none"
          height="32px"
          viewBox="0 0 32 32"
          width="32px"
        >
          <path
            d="M0.999999 1.38602L14.3404 14.7265C14.427 14.8132 14.5298 14.8819 14.643 14.9289C14.7562 14.9758 14.8775 14.9999 15 14.9999C15.1225 14.9999 15.2438 14.9758 15.357 14.9289C15.4702 14.8819 15.573 14.8132 15.6596 14.7265L29 1.38602"
            stroke="#2C232E"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.6"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;
